warehouse/query-core/src/main/java/datawave/query/jexl/nodes/QueryPropertyMarker.java

```diff
@@ -161,6 +165,59 @@ protected void setupSource(JexlNode source) {
         this.jjtAddChild(refExpNode1, 0);
     }
     
+    /**
+     * Determines if this source node is already marked with the provided marker class
+     *
+     * @param source
+     *            a JexlNode
+     * @param markerClass
+     *            an instance of a {@link QueryPropertyMarker}
+     * @return true if this node is already marked with the provided marker class
+     */
+    public static boolean isSourceMarked(JexlNode source, Class<? extends QueryPropertyMarker> markerClass) {
+        return QueryPropertyMarker.findInstance(source).isType(markerClass);
+    }
+    
+    /**
+     * Determine if one the source node's ancestors is already marked. This method does not check the source node for markers (see
+     * {@link #isSourceMarked(JexlNode, Class)}).
+     *
+     * Note: This method will recursively ascend the entire Jexl query tree to find a marked node. This imposes a non-zero cost for trees that are excessively
+     * deep or unflattened.
+     *
+     * @param node
+     *            a JexlNode
+     * @return true
+     */
+    public static boolean isAncestorMarked(JexlNode node, Class<? extends QueryPropertyMarker> markerClass) {
+        JexlNode parent = node.jjtGetParent();
+        if (parent != null) {
+            // marker nodes are by definition the intersection of a marker and a source
+            if (parent instanceof ASTAndNode && parent.jjtGetNumChildren() == 2) {
+                if (QueryPropertyMarker.findInstance(parent).isType(markerClass)) {
+                    return true;
+                }
+            }
+            return isAncestorMarked(parent, markerClass);
+        }
+        return false;
+    }
+    
+    /**
+     * Unwrap a marker node, fully. Intended to handle the odd edge case when multiple instances of the same marker is applied to the same node
+     *
+     * @param node
+     *            an arbitrary jexl node
+     * @return the source node, or the original node if the source node is not marked
+     */
+    public static JexlNode unwrapFully(JexlNode node, Class<? extends QueryPropertyMarker> marker) {
```

Comment:
This is sort of a placeholder method. It will unwrap a node tree for a specific marker type, for example, multiple delayed predicate markers. Once this feature is merged this method becomes obsolete since the same marker cannot be applied multiple times. 

If you are talking about unwrapping EVERY marker node from a given subtree, say unwrapping the source of a delayed bounded range, that would be something else entirely.